### DISCUSSION TITLE
Matplotlib demo for electrocardiogram

### DISCUSSION THREAD

**Discussion by vsquared** [py5coding/py5generator#486] [Show and tell]
I'm trying to use py5 for an app which uses a serial port connected to an Arduino UNO.  Imported data is then plotted using Processing's line() call.  However I get an error message that says "this.g" is null.  How can I initialize "this.g" in Python?  The same issue occurs with setting the background color.  Source code follows:
```
# Uses imported mode for py5
# Cannot invoke "processing.core.PGraphics.line(float, float, float, float)" because "this.g" is null
import serial

_wndW = 600
_wndH = 400

global x
x = 0
global y
y = 0
global y1
y1 = 0

def setup():
    size(_wndW, _wndH)
       
ser = serial.Serial('/dev/tty.usbmodem14301')  # open serial port   
print(ser.name)         # check which port was really used

  
while 1:    
    val = str(ser.readline())
    val = val.lstrip("b")
    val = val.strip("'")
    val = val.rstrip("\\r\\n")
    print(val)
    if len(val) > 0:
        y = int(float(val))
        
    line(x, _wndH-y-50, x-4, _wndH-y1-50)
    x = x+4
    y1 = y
    if (x > _wndW):
        x = 0
        background(0,0,255)
```
**Arduino code:**
```
byte outputValue = 0;

void setup() {
  Serial.begin(9600);
}

void loop() {
    outputValue += 4;
    Serial.println(outputValue);
}

```

**Comment by villares** [comment]
Cheers @vsquared!

If you are not using "static mode", in your case you are using "imported mode", but it would be the same in "module mode", you can't have a drawing method like `line()` outside `draw()` or some other py5-main-loop-controlled-functions.

So just replace `while 1:` by `def draw():` and believe it should work.

**Comment by vsquared** [comment]
It looks like 'while 1' is required to pull in the data but it seems to block anything else, eg, drawing into the window.  I added a separate draw() function similar to what I do in Processing but all I get is a rotating beachball; there is no drawing.

**Comment by hx2A** [comment]
@vsquared , for your first example, I would try creating a `draw()` function for drawing with py5 methods and then move the serial port stuff to its own function and have py5 call it repeatedly with `launch_repeating_thread()`. You wouldn't need the `while 1` loop to repeatedly process data from the Adruino and the drawing code would be taken care of by py5 in `draw()`.

https://py5coding.org/reference/sketch_launch_repeating_thread.html

**Reply by vsquared** [reply]
Thanks for the tip which does allow me to get a demo which works about as well as anything I've done in Processing.  I  prefer to draw from the data loop but this technique does not seem to be possible in Processing.  The demo plots most of the data, but some data points are lost somewhere between the data loop and draw().  Occasionally draw() just can't keep up and the plot becomes bizarre instead of a nice sawtooth pattern.  But overall it's generally pretty close. Note that I had to set the time delay to zero to get the data in realtime.   Below is the solution code for future reference:
```
import py5
import serial

_wndW = 600
_wndH = 400

global x,y,y1
x = 0
y = 0
y1 = 0

def settings():
    py5.size(_wndW, _wndH)

def setup():
    py5.launch_repeating_thread(get_data,name='get_data',time_delay=0)
   
def draw():
    global x,y,y1
    py5.line(x, _wndH-y-50, x-4, _wndH-y1-50)
    x = x+4
    y1 = y
    if (x > _wndW):
        x = 0
        py5.background(209)
       
ser = serial.Serial('/dev/tty.usbmodem14301')  # open serial port   
print(ser.name)         # check which port was really used

def get_data():
    global y
    
    val = str(ser.readline())
    val = val.lstrip("b")
    val = val.strip("'")
    val = val.rstrip("\\r\\n")
    print(val)
    if val != None:
        y = int(float(val))
        
py5.run_sketch()

```
**Output:**
<img width="604" alt="grph" src="https://github.com/user-attachments/assets/1d23d777-dfb5-4503-b684-8cb25d4f717f">

**GraphWin Output:**
Contrast the py5(Processing) output to that of Python GraphWin where the plotting was done in the while 1 loop.  The plot is much more pristine and no missing data points.

<img width="609" alt="grphWin" src="https://github.com/user-attachments/assets/ef7c08d3-8958-46fb-8903-5e36261f23da">

**Reply by hx2A** [reply]
> The demo plots most of the data, but some data points are lost somewhere between the data loop and draw(). Occasionally draw() just can't keep up and the plot becomes bizarre instead of a nice sawtooth pattern.

Have a look at the first matplotlib example here:

https://py5coding.org/integrations/matplotlib.html

It uses a deque to collect data that is later plotted. If your data loop added data to a deque, maybe with timestamps, your `draw()` function could then draw the data with nothing missing.

I recommend collecting timestamps because without it there might be temporal distortions in the plot.

**Reply by vsquared** [reply]
Thank you for the suggestions.  I'm not sure what a 'deque' is; I prefer that plotting be in realtime.  I have tried a large data array in the past with delayed plotting but don't use it much due to complexity.  I will take a look at matplotlib and appreciate your help.

**Reply by villares** [reply]
deque is a **d**ouble **e**nded **que**ue :)
A great easy high-level Python data structure to deal with a stream of data, or history, elegantly keeping a "window" of limited size, if you set a maximum length. It will then discard elements on the opposite side of your accretions.  https://docs.python.org/3/library/collections.html#collections.deque

**Reply by vsquared** [reply]
Thanks, very helpful.

**Reply by hx2A** [reply]
> A great easy high-level Python data structure to deal with a stream of data, or history, elegantly keeping a "window" of limited size, if you set a maximum length.

Well said! In Python the deque is also similar to lists with some extra functionality to cap the max length.

**Reply by vsquared** [reply]
The data points just keep getting scrunched up and the frame rate gets slower and slower (probably some setting I'm not aware of):
<img width="774" alt="4in1" src="https://github.com/user-attachments/assets/6fe18332-23fd-4c01-9633-28d25a116e77">

**Reply by hx2A** [reply]
Can we see the code? When you create the deque you should use the `maxlen` parameter to set the maximum length.

**Reply by vsquared** [reply]
plt.axis([0, None, 0, 300])
There is no top end (xmax); I want data points to be dropped off the bottom end when xmax exceeds the window width.
```python
import matplotlib.pyplot as plt
from matplotlib.animation import FuncAnimation
from matplotlib import style
import numpy as np
import random
import serial

global x,y

# initialize serial port
ser = serial.Serial()
ser.port = '/dev/cu.usbmodem14301' #Arduino serial port
ser.baudrate = 9600
# ser.timeout = 10 #specify timeout when using readline()
ser.open()
if ser.is_open==True:
    print("\nSerial port is open. Configuration:\n")
    print(ser, "\n") #print serial parameters

# Create figure for plotting
fig = plt.figure()
ax = fig.add_subplot(1, 1, 1)
xs = []
ys = []
x=0
y=0

# This function is called periodically from FuncAnimation
def animate(i, xs, ys):
    global x,y
    #Aquire and parse data from serial port
    val=str(ser.readline())      #ascii
    val = val.lstrip("b")
    val = val.strip("'")
    val = val.rstrip("\\r\\n")
#     print(val)
    print("["+str(x)+"]"+val)
    if val.isnumeric():
        val = int(val)
        ys.append(val)
        xs.append(x)
    x = x + 1
    print(ys[x-1])
    
    ax.clear()
    ax.plot(xs, ys)

    # Format plot
    plt.xticks(rotation=45, ha='right')
    plt.subplots_adjust(bottom=0.10)
    plt.title('Sawtooth Model')
    plt.ylabel('Value')
#     plt.legend()
# xmin,xmax,ymin,ymax
    plt.axis([0, None, 0, 300]) #Use for arbitrary number of trials

animation = FuncAnimation(fig, animate, fargs=(xs,ys), interval=1)

plt.show()

```

**Reply by vsquared** [reply]
I figured out how to show just a certain amount of data points:
plt.axis has to be set to ([None, None, 0, 300]) and you have to use deque's for the data points with the maxlen set, eg
```
xs = deque([],maxlen=550)
ys = deque([],maxlen=550)
```
I'll edit the above post later.

**Reply by hx2A** [reply]
Yes, that's a good use of the deque data structure!

**Comment by vsquared** [comment]
The following Python matplotlib source code provides an alternative to the initial post by creating a scrolling data plot when connected to an Arduino UNO through a serial port (Arduino code shown in original post).  By using two _deque_ arrays for the x and y data points a limit of 550 data points are plotted.  When the window is resized only 550 data points are shown at each size.  Initially the x-axis is static but starts scrolling after 550 data points are plotted.  By setting _maxlen_ in each of the deque arrays, data points are removed from the left side of the plot as new data points are added to the right side.  This creates a 'scrolling' graph that appears to be slower than the incoming data.   However, there are no obvious missing data points as seen with the Processing model.

```python
import matplotlib.pyplot as plt
from matplotlib.animation import FuncAnimation
from matplotlib import style
from collections import deque

import numpy as np
import random
import serial

global x,y

# initialize serial port
ser = serial.Serial()
ser.port = '/dev/cu.usbmodem14301' #Arduino serial port
ser.baudrate = 9600
ser.open()

if ser.is_open==True:
    print("\nSerial port is open. Configuration:\n")
    print(ser, "\n") #print serial parameters

# Create figure for plotting
fig = plt.figure()
ax = fig.add_subplot(1, 1, 1)

xs = deque([],maxlen=550)
ys = deque([],maxlen=550)
x=0
y=0

# This function is called periodically from FuncAnimation
def animate(i, xs, ys):
    global x,y
    #Aquire and parse data from serial port
    val=str(ser.readline())      #ascii
    val = val.lstrip("b")
    val = val.strip("'")
    val = val.rstrip("\\r\\n")
    print("["+str(x)+"]"+val)
    if val.isnumeric():
        val = int(val)
        ys.append(val)
        xs.append(x)
    x = x + 1
       
    ax.clear()
    ax.plot(xs, ys)

    # Format plot
    plt.xticks(rotation=45, ha='right')
    plt.subplots_adjust(bottom=0.10)
    plt.title('Sawtooth Model')
    plt.ylabel('Value')

# xmin,xmax,ymin,ymax
    plt.axis([None, None, 0, 300]) #Use for arbitrary number of trials

animation = FuncAnimation(fig, animate, fargs=(xs,ys), interval=1)

plt.show()

```
**Data limit is honored at all window sizes:**
<img width="768" alt="limit" src="https://github.com/user-attachments/assets/270c3cf8-f0ee-44b2-a71c-66506ecbedaa">

**Addendum:**
...And it can do something besides sawtooth patterns:

<img width="653" alt="ecg" src="https://github.com/user-attachments/assets/fd9101fd-5b5d-40f5-a975-ac8d419b0afd">

**Reply by villares** [reply]
Cheers @vsquared, wouldn't you like to rename this discussion thread to something that describes what you achieved? I moved it to "show and tell" already :D

**Reply by vsquared** [reply]
Any way to split it?  The first part provided a work around to the problem since we're unable to use line() outside of draw()(similar to Processing).  The second part followed the suggestion to use matplotlib which I had not used before.

**Reply by villares** [reply]
Good idea! 

I have opened another discussion, about `Cannot invoke "processing.core.PGraphics.line(float, float, float, float)" because "this.g" is null` at https://github.com/py5coding/py5generator/discussions/487 

Now you can rename this!

**Reply by vsquared** [reply]
I don't know what else to call it.  This thread started off with a Processing error message (which I have seen before) and wound up with a matplotlib demo that can be used for plotting an electrocardiogram.  A lot of territory was covered in this one thread and I learned quite a bit.  For some reason I thought I might be able to use a drawing call in py5 like I've done with GraphWin (inside a _while_ construct), but it turns out I was wrong and py5 has the same limitations as Processing when it comes to drawing graphics (has to be done inside draw()).

**Reply by hx2A** [reply]
> py5 has the same limitations as Processing when it comes to drawing graphics (has to be done inside draw()).

Well, there are ways to use py5 without a `draw()` or `setup()` function, but they don't open an animated window. For example:

https://py5coding.org/reference/py5functions_render.html

I don't think that's the right fit for this situation but it probably would be possible to do something else that was, particularly if it was limited to the default JAVA2D renderer.

**Reply by villares** [reply]
> matplotlib demo that can be used for plotting an electrocardiogram.

This could be a good name!

It doesn't bother me at all using draw() as the main animation loop. Also, sometimes I have a separate thread at a different speed updating a data structure that I can "sample" at draw time.

**Reply by vsquared** [reply]
The matplotlib demo doesn't use py5; is that a problem?
` I have a separate thread at a different speed updating a data structure`
That's why Processing has _void serialEvent(Serial p) {...}_ but I don't think we can use that in py5.

**Reply by villares** [reply]
No problem. If you add py5 to your example afterwards you can update it.

have a look at the easy threading support by py5: https://py5coding.org/reference/#threading maybe the `launch_repeating_thread()`, as Jim mentioned... or `launch_thread()` with your while loop inside it...

**Reply by vsquared** [reply]
`have a look at the easy threading support`

I already tried that and the output left a little to be desired.  There are images in this thread.  Losing data points is not good for my purposes.
